// discord-player-youtube/dist/internal/youtube.js
import ytdl from "./ytdl.js";
import { toSecond, parseNumber, clone, toCookieArray, } from "./utils.js";
import { ytsrSearch } from "./ytsr.js";
import { ytplResolve } from "./ytpl.js";
import { ytjsResolveMix } from "./ytjs.js";
import { extractPlaylistIdFromUrl } from "./utils.js";
let _cookies;
export function configure(opts = {}) {
    _cookies = toCookieArray(opts.cookies);
    ytdl.init({ cookies: _cookies, ytdlOptions: opts.ytdlOptions ? clone(opts.ytdlOptions) : {} });
}
function ensureAgent() {
    ytdl.ensure();
}
function mergedYtdlOptions() {
    return ytdl.options();
}
export async function search(query, options = {}) {
    ensureAgent();
    const res = await ytsrSearch(query, options);
    return res ?? { playlist: null, items: [] };
}
export async function resolve(url) {
    ensureAgent();
    // Delegate playlists to appropriate resolver: ytjs for mixes, ytpl for regular playlists
    try {
        const playlistId = extractPlaylistIdFromUrl(url);
        if (playlistId) {
            const isMix = /^RD/i.test(String(playlistId));
            if (isMix) {
                const mix = await ytjsResolveMix(url, playlistId);
                if (mix)
                    return mix;
            }
            else {
                const pl = await ytplResolve(url);
                if (pl)
                    return pl;
            }
        }
    }
    catch { }
    // Canonicalize to a proper watch URL if needed
    let target = url;
    if (!ytdl.validateURL(target)) {
        try {
            const vid = ytdl.getVideoID(target);
            target = `https://youtu.be/${vid}`;
        }
        catch {
            // Fall back to search
        }
    }
    if (ytdl.validateURL(target)) {
        let info;
        try {
            info = await ytdl.getBasicInfo(target, mergedYtdlOptions());
        }
        catch (e) {
            // Fallback try WEB client only (preserve agent)
            const base = { ...mergedYtdlOptions() };
            base.playerClients = ["WEB"];
            info = await ytdl.getBasicInfo(target, base);
        }
        const d = info.videoDetails;
        const item = {
            source: "youtube",
            url: d.video_url || `https://youtu.be/${d.videoId}`,
            title: d.title ?? "Unknown",
            author: d.author?.name || d.author?.user || "Unknown",
            durationMS: (d.isLive ? 0 : toSecond(d.lengthSeconds)) * 1000,
            thumbnail: d.thumbnails?.sort((a, b) => b.width - a.width)?.[0]?.url,
            isLive: Boolean(d.isLive),
            views: parseNumber(d.viewCount || d.view_count || d.views),
            likes: parseNumber(d.likes),
            raw: {
                chapters: d.chapters ?? [],
                storyboards: d.storyboards ?? [],
                related: info.related_videos ?? [],
                info,
            },
        };
        return { playlist: null, items: [item] };
    }
    // Fallback to search
    return search(url);
}
export async function related(url) {
    ensureAgent();
    if (!ytdl.validateURL(url))
        return search(url);
    const basic = await ytdl.getBasicInfo(url, mergedYtdlOptions());
    const related = (basic.related_videos?.filter((r) => r.id) ?? [])
        // Filter out ultra-short related results (likely Shorts/clips)
        .filter((r) => {
        const len = Number(r.length_seconds) || 0;
        return len === 0 || len >= 60; // keep if unknown or >= 60s
    });
    const items = related.map((r) => ({
        source: "youtube",
        url: `https://youtu.be/${r.id}`,
        title: r.title ?? "Unknown",
        author: typeof r.author === "string"
            ? r.author
            : r.author?.name || r.author?.user || "Unknown",
        durationMS: (r.isLive ? 0 : toSecond(r.length_seconds)) * 1000,
        thumbnail: r.thumbnails?.sort((a, b) => b.width - a.width)?.[0]?.url,
        isLive: Boolean(r.isLive),
        views: parseNumber(r.view_count),
        raw: r,
    }));
    return { playlist: null, items };
}
export async function streamUrl(url) {
    ensureAgent();
    let target = url;
    if (!ytdl.validateURL(target)) {
        try {
            const vid = ytdl.getVideoID(target);
            target = `https://youtu.be/${vid}`;
        }
        catch {
            throw new Error("CANNOT_RESOLVE_SONG");
        }
    }
    let info;
    try {
        info = await ytdl.getInfo(target, mergedYtdlOptions());
    }
    catch (e) {
        // Fallback: try WEB client only (some tokens/regions prefer WEB over TV)
        try {
            const base = { ...mergedYtdlOptions() };
            base.playerClients = ["WEB"];
            info = await ytdl.getInfo(target, base);
        }
        catch {
            throw e;
        }
    }
    if (!info.formats?.length)
        throw new Error("UNAVAILABLE_VIDEO");
    const details = info.videoDetails;
    const isLive = Boolean(details?.isLive);
    const format = info.formats
        .filter((f) => f.hasAudio && (!isLive || f.isHLS))
        .sort((a, b) => Number(b.audioBitrate) - Number(a.audioBitrate) || Number(a.bitrate) - Number(b.bitrate))[0];
    if (!format)
        throw new Error("UNPLAYABLE_FORMATS");
    return format.url;
}
