// discord-player-youtube/dist/internal/ytdl.js
import core from "@distube/ytdl-core";
import { clone, toSecond, parseNumber } from "./utils.js";
let _cookies;
let _options = {};
let _agent;
export function configure(opts = {}) {
    _cookies = opts.cookies ? clone(opts.cookies) : undefined;
    _options = opts.ytdlOptions ? clone(opts.ytdlOptions) : {};
    _agent = core.createAgent(_cookies);
    _options.agent = _agent;
}
export function ensureAgent() {
    if (!_options.agent) {
        _agent = core.createAgent(_cookies);
        _options.agent = _agent;
    }
}
export function cookieHeader() {
    const agent = _options.agent;
    if (!agent?.jar)
        return "";
    return agent.jar.getCookieStringSync?.("https://www.youtube.com") ?? "";
}
export function authHeaders() {
    const headers = {};
    const cookie = cookieHeader();
    if (cookie)
        headers["cookie"] = cookie;
    return headers;
}
export function mergedOptions() {
    const base = { ..._options };
    const h = authHeaders();
    base.requestOptions = { ...(base.requestOptions || {}), headers: { ...(base.requestOptions?.headers || {}), ...h } };
    if (_options.agent)
        base.agent = _options.agent;
    return base;
}
export const validateURL = core.validateURL;
export const getVideoID = core.getVideoID;
export async function getBasicInfo(url, opts) {
    return core.getBasicInfo(url, opts ?? mergedOptions());
}
export async function getInfo(url, opts) {
    return core.getInfo(url, opts ?? mergedOptions());
}
export async function streamUrl(url) {
    let target = url;
    if (!core.validateURL(target)) {
        try {
            const vid = core.getVideoID(target);
            target = `https://youtu.be/${vid}`;
        }
        catch {
            throw new Error('CANNOT_RESOLVE_SONG');
        }
    }
    let info;
    try {
        info = await core.getInfo(target, mergedOptions());
    }
    catch (e) {
        try {
            const base = { ...mergedOptions() };
            base.playerClients = ['WEB'];
            info = await core.getInfo(target, base);
        }
        catch {
            throw e;
        }
    }
    if (!info.formats?.length)
        throw new Error('UNAVAILABLE_VIDEO');
    const details = info.videoDetails;
    const isLive = Boolean(details?.isLive);
    const format = info.formats
        .filter((f) => f.hasAudio && (!isLive || f.isHLS))
        .sort((a, b) => Number(b.audioBitrate) - Number(a.audioBitrate) || Number(a.bitrate) - Number(b.bitrate))[0];
    if (!format)
        throw new Error('UNPLAYABLE_FORMATS');
    return String(format.url);
}
export async function ytdlResolveVideo(url) {
    try {
        let target = url;
        if (!validateURL(target)) {
            try {
                const vid = getVideoID(target);
                target = `https://youtu.be/${vid}`;
            }
            catch { }
        }
        let info;
        try {
            info = await getBasicInfo(target, mergedOptions());
        }
        catch (e) {
            const base = { ...mergedOptions() };
            base.playerClients = ['WEB'];
            info = await getBasicInfo(target, base);
        }
        const d = info.videoDetails;
        const item = {
            source: "youtube",
            url: d.video_url || `https://youtu.be/${d.videoId}`,
            title: d.title ?? "Unknown",
            author: d.author?.name || d.author?.user || "Unknown",
            durationMS: (d.isLive ? 0 : toSecond(d.lengthSeconds)) * 1000,
            thumbnail: d.thumbnails?.sort((a, b) => b.width - a.width)?.[0]?.url,
            isLive: Boolean(d.isLive),
            views: parseNumber(d.viewCount || d.view_count || d.views),
            likes: parseNumber(d.likes),
            raw: { chapters: d.chapters ?? [], storyboards: d.storyboards ?? [], related: info.related_videos ?? [], info }
        };
        return { playlist: null, items: [item] };
    }
    catch {
        return null;
    }
}
// Convenience API to keep callers clean
export const api = {
    init: configure,
    ensure: ensureAgent,
    cookie: cookieHeader,
    headers: authHeaders,
    options: mergedOptions,
    validateURL,
    getVideoID,
    getBasicInfo,
    getInfo,
    streamUrl,
    ytdlResolveVideo,
};
export default api;
