// discord-player-youtube/dist/internal/ytjs.js
let _clientPromise = null;
let _disabled = false;
let _authToken;
let _cookieString;
let _poToken;
let _wantPoToken = false;
let _innertubeConfigRaw;
import Innertube, { UniversalCache, ClientType, YTNodes } from 'youtubei.js';
import { msFromDurationLike, parseNumber, tokenToObject, generatePoToken, generatePoTokenForInnertube, } from './utils.js';
export function configureYTJS(opts = {}) {
    _authToken =
        typeof opts.authentication === 'string' &&
            opts.authentication.trim().length > 0
            ? opts.authentication.trim()
            : undefined;
    _cookieString = toCookieHeader(opts.cookie) ?? _cookieString;
    _innertubeConfigRaw = opts.innertubeConfigRaw;
    _wantPoToken = Boolean(opts.generateWithPoToken);
    if (_wantPoToken && !_poToken) {
        try {
            _poToken = generatePoToken();
        }
        catch {
            /* ignore */
        }
    }
}
async function createClient(init) {
    try {
        const InnertubeCtor = Innertube;
        if (!InnertubeCtor)
            return null;
        const payload = {};
        const tokenString = _authToken;
        const cacheDir = process?.env?.YT_CACHE_DIR;
        const configuredCookie = cookieHeader();
        const rawConfig = init?.innertubeConfigRaw ?? _innertubeConfigRaw;
        const createInnertube = rawConfig != null
            ? (payload) => InnertubeCtor.create(payload, rawConfig)
            : (payload) => InnertubeCtor.create(payload);
        if (tokenString) {
            // Prefer TV OAuth client when tokens are provided
            if (ClientType)
                payload.client_type = ClientType.TV_EMBEDDED;
            if (UniversalCache)
                payload.cache = new UniversalCache(true, cacheDir);
            if (configuredCookie)
                payload.cookie = configuredCookie;
            if (_poToken)
                payload.po_token = _poToken;
            const client = await createInnertube(payload);
            try {
                const tokens = tokenToObject(tokenString);
                await client.session.signIn(tokens);
                // Best-effort info logging (avoid bringing a logger dependency)
                try {
                    const info = await client.account.getInfo();
                    // eslint-disable-next-line no-console
                    console.info(info.contents?.contents
                        ? `[discord-player-youtube] Signed into YouTube using the name: ${info.contents.contents[0].is(YTNodes.AccountItem)
                            ? info.contents.contents[0].as(YTNodes.AccountItem)
                                .account_name.text ?? 'UNKNOWN ACCOUNT'
                            : 'UNKNOWN ACCOUNT'}`
                        : `[discord-player-youtube] Signed into YouTube using the client name: ${client.session.client_name}@${client.yt.session.client_version}`);
                }
                catch { }
            }
            catch (err) {
                // If token parsing/sign-in fails, continue with unsigned client
                try {
                    // eslint-disable-next-line no-console
                    console.warn(`YouTubeExtractor: TV OAuth sign-in failed; continuing unsigned. Reason: ${String(err?.message || err)}`);
                }
                catch { }
            }
            // Optionally improve PoToken after client is available (has visitorData)
            // Always attempt improvement when requested; replace fallback token if a better one is generated.
            if (_wantPoToken) {
                try {
                    const pot = await generatePoTokenForInnertube(client);
                    if (pot && typeof pot === 'string') {
                        _poToken = pot;
                        try {
                            if (client?.session) {
                                client.session.po_token = pot;
                                if (client.session.player)
                                    client.session.player.po_token = pot;
                            }
                        }
                        catch { }
                    }
                }
                catch { }
            }
            return client;
        }
        // Cookie-based or anonymous client
        if (init?.cookie)
            payload.cookie = init.cookie;
        else if (configuredCookie)
            payload.cookie = configuredCookie;
        if (init?.options && typeof init.options === 'object')
            Object.assign(payload, init.options);
        if (_poToken)
            payload.po_token = _poToken;
        const client = await createInnertube(payload);
        if (_wantPoToken) {
            try {
                const pot = await generatePoTokenForInnertube(client);
                if (pot && typeof pot === 'string') {
                    _poToken = pot;
                    try {
                        if (client?.session) {
                            client.session.po_token = pot;
                            if (client.session.player)
                                client.session.player.po_token = pot;
                        }
                    }
                    catch { }
                }
            }
            catch { }
        }
        return client;
    }
    catch {
        return null;
    }
}
export async function getYTJSClient(init = {}) {
    if (_disabled)
        return null;
    if (_clientPromise)
        return _clientPromise;
    _clientPromise = (async () => {
        const client = await createClient(init);
        if (!client)
            _disabled = true;
        return client;
    })();
    return _clientPromise;
}
export function disableYTJS() {
    _disabled = true;
}
export function isYTJSEnabled() {
    return !_disabled;
}
export async function streamUrl(url, cookie, useClient) {
    const vid = extractVideoIdFromAny(url);
    const client = await getYTJSClient({
        cookie: cookie ?? cookieHeader(),
        options: {},
    });
    if (!client || !vid)
        throw new Error('YTJS_UNAVAILABLE');
    // Preferred: youtubei.js-deciphered URL
    try {
        const fmt = await client.getStreamingData(vid, {
            type: 'audio',
            quality: 'best',
            client: useClient || 'ANDROID',
        });
        if (fmt?.url)
            return String(fmt.url);
    }
    catch {
        /* ignore and try fallback */
    }
    // Fallback: chooseFormat + decipher
    try {
        const info = await client.getInfo(vid, {
            client: useClient || 'ANDROID',
        });
        if (typeof info?.chooseFormat === 'function') {
            const fmt = info.chooseFormat({ type: 'audio', quality: 'best' });
            const u = typeof fmt?.decipher === 'function'
                ? fmt.decipher(client.session?.player)
                : fmt?.url;
            if (u)
                return String(u);
        }
        // Last resort: raw URL from streaming_data (may fail for ciphered entries)
        const sd = info.streaming_data ||
            info.player_response?.streamingData;
        const formats = []
            .concat(sd?.adaptive_formats || [])
            .concat(sd?.formats || []);
        const audioOnly = formats.filter((f) => (!f.height && !f.width) ||
            f?.vcodec === 'none' ||
            (f?.has_audio && !f?.has_video));
        const best = audioOnly.sort((a, b) => Number(b?.bitrate || b?.audioBitrate || 0) -
            Number(a?.bitrate || a?.audioBitrate || 0))[0];
        if (best?.url)
            return String(best.url);
    }
    catch {
        /* ignore */
    }
    throw new Error('YTJS_NO_STREAM');
}
// ===== Helpers specific to youtubei.js structures
function textFromRuns(obj) {
    try {
        const runs = obj?.runs;
        if (runs?.length)
            return String(runs.map((r) => r.text).join(''));
        const simple = obj?.simpleText;
        if (simple)
            return String(simple);
    }
    catch { }
    return undefined;
}
function thumbFromThumbs(t) {
    try {
        const thumbs = t?.thumbnails;
        if (Array.isArray(thumbs) && thumbs.length)
            return String(thumbs.sort((a, b) => (b.width || 0) - (a.width || 0))[0].url);
    }
    catch { }
    return undefined;
}
// Local cookie + URL helpers (use only cookie provided at registration)
function toCookieHeader(input) {
    try {
        if (!input)
            return undefined;
        if (typeof input === 'string')
            return input.trim() || undefined;
        // Tough-cookie CookieJar support
        if (typeof input?.getCookieStringSync === 'function') {
            const v = input.getCookieStringSync('https://www.youtube.com');
            if (typeof v === 'string' && v.trim())
                return v.trim();
        }
        if (typeof input?.jar?.getCookieStringSync === 'function') {
            const v = input.jar.getCookieStringSync('https://www.youtube.com');
            if (typeof v === 'string' && v.trim())
                return v.trim();
        }
        // Array of cookie-like objects
        if (Array.isArray(input)) {
            const parts = input
                .map((c) => {
                if (!c)
                    return '';
                const name = (c.name ?? c.key ?? '').toString();
                const value = (c.value ?? '').toString();
                if (name && value)
                    return `${name}=${value}`;
                if (typeof c.toString === 'function') {
                    const s = String(c.toString());
                    if (s.includes('='))
                        return s;
                }
                return '';
            })
                .filter(Boolean);
            const joined = parts.join('; ').trim();
            return joined || undefined;
        }
        // Object map of name -> value
        if (typeof input === 'object') {
            const parts = Object.entries(input)
                .map(([k, v]) => (k && v ? `${k}=${v}` : ''))
                .filter(Boolean);
            const joined = parts.join('; ').trim();
            return joined || undefined;
        }
    }
    catch { }
    return undefined;
}
function cookieHeader() {
    return _cookieString;
}
function extractVideoIdFromAny(href) {
    // Try robust URL parsing first
    try {
        const u = new URL(href);
        // Short links: https://youtu.be/<id>
        if (/(^|\.)youtu\.be$/i.test(u.hostname)) {
            const seg = u.pathname.split('/').filter(Boolean)[0];
            if (seg && seg.length >= 8)
                return seg;
        }
        // Watch links: https://www.youtube.com/watch?v=<id>
        const v = u.searchParams.get('v');
        if (v && v.trim().length >= 8)
            return v.trim();
        // Shorts: https://www.youtube.com/shorts/<id>
        const m = u.pathname.match(/\/shorts\/([a-zA-Z0-9_-]{8,})/);
        if (m)
            return m[1];
    }
    catch { }
    // Fallback: loose ID pattern (best-effort)
    const m = href.match(/([a-zA-Z0-9_-]{11})/);
    return m ? m[1] : null;
}
// ===== Provider-specific exported functions (moved from utils)
export async function ytjsSearch(query, options) {
    const client = await getYTJSClient({ cookie: cookieHeader() });
    if (!client)
        return null;
    try {
        const limit = Math.min(options?.limit ?? 25, 25);
        let search = await client.search(query);
        let vids = Array.isArray(search?.videos) ? [...search.videos] : [];
        // If fewer than requested, try to fetch continuation pages
        while (vids.length < limit) {
            try {
                if (typeof search?.getContinuation !== 'function')
                    break;
                const next = await search.getContinuation();
                if (!next)
                    break;
                search = next;
                const more = Array.isArray(search?.videos) ? search.videos : [];
                if (!more.length)
                    break;
                vids = vids.concat(more);
            }
            catch {
                break;
            }
        }
        const items = vids.slice(0, limit).map((v) => {
            const title = typeof v.title?.toString === 'function'
                ? v.title.toString()
                : v.title || 'Unknown';
            const author = v.author?.name ||
                (typeof v.author?.toString === 'function'
                    ? v.author.toString()
                    : 'Unknown');
            const durationLike = typeof v.length_text?.toString === 'function'
                ? v.length_text.toString()
                : v.duration?.toString?.() || '0:00';
            const thumb = Array.isArray(v.thumbnails) && v.thumbnails.length
                ? String(v.thumbnails.sort((a, b) => (b.width || 0) - (a.width || 0))[0].url)
                : undefined;
            return {
                source: 'youtube',
                url: `https://youtu.be/${v.video_id || v.id}`,
                title: title || 'Unknown',
                author: author || 'Unknown',
                durationMS: msFromDurationLike(durationLike),
                thumbnail: thumb,
                isLive: !v.length_text && !v.duration,
                views: parseNumber(v.view_count),
                raw: v,
            };
        });
        return { playlist: null, items };
    }
    catch {
        return null;
    }
}
export async function ytjsResolvePlaylist(playlistId, url) {
    const client = await getYTJSClient({ cookie: cookieHeader() });
    if (!client)
        return null;
    try {
        const resp = await client.actions.execute('/browse', {
            browseId: `VL${playlistId}`,
        });
        const root = resp?.data || {};
        const plroot = root?.contents?.twoColumnBrowseResultsRenderer?.tabs?.[0]?.tabRenderer
            ?.content?.sectionListRenderer?.contents?.[0]?.itemSectionRenderer
            ?.contents?.[0]?.playlistVideoListRenderer;
        const contents = plroot?.contents || [];
        const items = [];
        for (const c of contents) {
            const v = c.playlistVideoRenderer;
            if (!v?.videoId)
                continue;
            const title = textFromRuns(v.title) || 'Unknown';
            const author = textFromRuns(v.shortBylineText) ||
                textFromRuns(v.longBylineText) ||
                'Unknown';
            const duration = textFromRuns(v.lengthText) || '0:00';
            const thumbnail = thumbFromThumbs(v.thumbnail);
            items.push({
                source: 'youtube',
                url: `https://youtu.be/${v.videoId}`,
                title,
                author,
                durationMS: msFromDurationLike(duration),
                thumbnail,
                isLive: !v.lengthText,
                playlistId,
                raw: v,
            });
        }
        if (!items.length)
            return null;
        return {
            playlist: {
                id: playlistId,
                title: undefined,
                url: `https://www.youtube.com/playlist?list=${playlistId}`,
                thumbnail: items[0]?.thumbnail,
            },
            items,
        };
    }
    catch {
        return null;
    }
}
export async function ytjsResolveMix(url, playlistId) {
    const client = await getYTJSClient({ cookie: cookieHeader() });
    if (!client)
        return null;
    const videoId = extractVideoIdFromAny(url);
    if (!videoId)
        return null;
    try {
        const resp = await client.actions.execute('/next', {
            videoId,
            playlistId,
        });
        const root = resp?.data || {};
        const playlist = root?.contents?.twoColumnWatchNextResults?.playlist?.playlist;
        const contents = playlist?.contents || [];
        const items = [];
        for (const c of contents) {
            const v = c.playlistPanelVideoRenderer;
            if (!v?.videoId)
                continue;
            const title = textFromRuns(v.title) || 'Unknown';
            const author = textFromRuns(v.shortBylineText) ||
                textFromRuns(v.longBylineText) ||
                'Unknown';
            const duration = textFromRuns(v.lengthText) || '0:00';
            const thumbnail = thumbFromThumbs(v.thumbnail);
            items.push({
                source: 'youtube',
                url: `https://youtu.be/${v.videoId}`,
                title,
                author,
                durationMS: msFromDurationLike(duration),
                thumbnail,
                isLive: !v.lengthText,
                playlistId,
                playlistTitle: textFromRuns(playlist?.title) || 'YouTube Mix',
                raw: v,
            });
        }
        if (!items.length)
            return null;
        return {
            playlist: {
                id: playlistId,
                title: textFromRuns(playlist?.title) || 'YouTube Mix',
                url: `https://youtu.be/${videoId}&list=${playlistId}`,
                thumbnail: items[0]?.thumbnail,
            },
            items,
        };
    }
    catch {
        return null;
    }
}
export async function ytjsResolveVideo(url) {
    const client = await getYTJSClient({ cookie: cookieHeader() });
    if (!client)
        return null;
    try {
        const vid = extractVideoIdFromAny(url);
        if (!vid)
            return null;
        const info = await client.getInfo(vid);
        const b = info?.basic_info || {};
        const thumb = Array.isArray(b.thumbnail?.thumbnails) && b.thumbnail.thumbnails.length
            ? String(b.thumbnail.thumbnails.sort((a, b) => (b.width || 0) - (a.width || 0))[0].url)
            : undefined;
        const item = {
            source: 'youtube',
            url: `https://youtu.be/${b.id}`,
            title: b.title ?? 'Unknown',
            author: typeof b.author === 'string' ? b.author : b.channel?.name || 'Unknown',
            durationMS: msFromDurationLike(String(b.duration ?? 0)),
            thumbnail: thumb,
            isLive: Boolean(b.is_live),
            views: parseNumber(b.view_count),
            likes: parseNumber(b.like_count),
            raw: { info },
        };
        return { playlist: null, items: [item] };
    }
    catch {
        return null;
    }
}
